File: scripts/verify-liblib-batch80.mjs
import assert from "node:assert/strict";
import { createDirectorProjectDocumentV1 } from "../src/lib/directorProjectDocument.ts";
import {
  createDirectorProjectStorageKey,
  DirectorProjectPersistenceAuthority,
} from "../src/lib/directorProjectPersistence.ts";

class MemoryBackend {
  values = new Map();
  failWrites = false;

  getItem(key) {
    return this.values.get(key) ?? null;
  }

  setItem(key, value) {
    if (this.failWrites) throw new Error("quota");
    this.values.set(key, value);
  }

  removeItem(key) {
    this.values.delete(key);
  }
}

const owner = {
  route: "libtv",
  canvasId: "batch80-canvas-a",
  sourceNodeId: "batch80-director-a",
};

const transform = (position = [0, 0, 0]) => ({
  position,
  rotation: [0, 0, 0],
  scale: [1, 1, 1],
});

function createDocument(name = "Batch 80 scene") {
  const camera = {
    id: "batch80-camera",
    name: "Main camera",
    kind: "camera",
    primitive: "camera",
    color: "#99ddff",
    visible: true,
    locked: false,
    transform: transform([4, 2, 6]),
    camera: {
      fov: 45,
      target: [0, 1, 0],
      lookAtMode: "coordinate",
      lookAtObjectId: null,
      followTargetId: null,
      followOffset: [0, 1.2, 4.5],
      followView: "third-person",
    },
  };
  return createDirectorProjectDocumentV1({
    projectId: "batch80-project-a",
    owner,
    scene: {
      name,
      backgroundColor: "#20252b",
      groundColor: "#30343a",
      showGround: true,
      showGrid: true,
    },
    objects: [camera],
    groups: [],
    activeCameraId: camera.id,
    aspectRatio: "16:9",
    timeline: {
      duration: 8,
      currentTime: 4,
      isPlaying: true,
      loop: true,
      zoom: 1,
      autoKeyframe: true,
      tracks: [],
      motionPaths: [],
      selectedTrackId: null,
      selectedKeyframeId: null,
      selectedMotionPathId: null,
      selectedMotionPathAnchorId: null,
      selectedMotionPathHandle: null,
      motionPathDraft: null,
      editorMode: "timeline",
      cameraMotionPreset: { application: null, error: null },
    },
    captures: [
      {
        id: "batch80-capture",
        dataUrl: "data:image/png;base64,TRANSIENT_BYTES",
        cameraId: camera.id,
        cameraName: camera.name,
        aspectRatio: "16:9",
        width: 1280,
        height: 720,
        createdAt: "2026-08-28T12:00:00.000Z",
        sentNodeId: null,
      },
    ],
  });
}

function rawValue(backend, key) {
  const raw = backend.getItem(key);
  assert.ok(raw);
  return JSON.parse(raw);
}

const backend = new MemoryBackend();
const authority = new DirectorProjectPersistenceAuthority(backend);
const document = createDocument();
const key = createDirectorProjectStorageKey(owner);
const saved = authority.save({
  owner,
  projectId: document.projectId,
  generation: 3,
  document,
  savedAt: "2026-08-28T12:01:00.000Z",
});
assert.equal(saved.disposition, "SAVED");
assert.equal(saved.envelope?.document.captureDescriptors.length, 0);

const tombstoned = authority.tombstone({
  owner,
  projectId: document.projectId,
  generation: 3,
  tombstonedAt: "2026-08-28T12:02:00.000Z",
});
assert.equal(tombstoned.disposition, "TOMBSTONED");
assert.deepEqual(rawValue(backend, key), {
  storageSchemaVersion: 1,
  lifecycle: "TOMBSTONED",
  projectId: document.projectId,
  owner,
  generation: 3,
  tombstonedAt: "2026-08-28T12:02:00.000Z",
});
assert.equal(authority.load(owner).reason, "PROJECT_TOMBSTONED");

const repeated = authority.tombstone({
  owner,
  projectId: document.projectId,
  generation: 3,
});
assert.equal(repeated.disposition, "ALREADY_TOMBSTONED");
assert.equal(repeated.envelope?.tombstonedAt, "2026-08-28T12:02:00.000Z");

const saveAfterTombstone = authority.save({
  owner,
  projectId: document.projectId,
  generation: 4,
  document: createDocument("must not resurrect"),
});
assert.equal(saveAfterTombstone.disposition, "STALE_IGNORED");
assert.equal(saveAfterTombstone.reason, "PROJECT_TOMBSTONED");
assert.equal(rawValue(backend, key).lifecycle, "TOMBSTONED");

const malformed = {
  ...rawValue(backend, key),
  unexpected: true,
};
backend.setItem(key, JSON.stringify(malformed));
const malformedResult = authority.load(owner);
assert.equal(malformedResult.disposition, "REJECTED");
assert.equal(malformedResult.reason, "INVALID_ENVELOPE");

const staleBackend = new MemoryBackend();
const staleAuthority = new DirectorProjectPersistenceAuthority(staleBackend);
staleAuthority.save({
  owner,
  projectId: document.projectId,
  generation: 9,
  document,
});
const staleTombstone = staleAuthority.tombstone({
  owner,
  projectId: document.projectId,
  generation: 8,
});
assert.equal(staleTombstone.disposition, "REJECTED");
assert.equal(staleTombstone.reason, "STALE_REQUEST");
assert.equal(staleAuthority.load(owner).disposition, "RESTORED");

const failedBackend = new MemoryBackend();
const failedAuthority = new DirectorProjectPersistenceAuthority(failedBackend);
failedAuthority.save({
  owner,
  projectId: document.projectId,
  generation: 1,
  document,
});
failedBackend.failWrites = true;
const failedTombstone = failedAuthority.tombstone({
  owner,
  projectId: document.projectId,
  generation: 1,
});
assert.equal(failedTombstone.disposition, "REJECTED");
assert.equal(failedTombstone.reason, "WRITE_FAILED");
assert.equal(failedAuthority.getRecord(owner).status, "SESSION_ONLY");
assert.equal(failedAuthority.load(owner).disposition, "RESTORED");

const unavailable = new DirectorProjectPersistenceAuthority(null);
assert.equal(
  unavailable.tombstone({
    owner,
    projectId: document.projectId,
    generation: 1,
  }).reason,
  "STORAGE_UNAVAILABLE",
);

console.log(
  JSON.stringify(
    {
      status: "PASS",
      batch: 80,
      scenarios: [
        "strict tombstone envelope round-trip",
        "durable tombstone blocks load",
        "tombstone is idempotent",
        "save cannot resurrect tombstoned project",
        "malformed tombstone is rejected",
        "stale tombstone cannot replace newer document",
        "write failure preserves session-only continuity",
        "storage unavailable",
      ],
      errors: [],
    },
    null,
    2,
  ),
);

File: scripts/verify-liblib-batch80.py
#!/usr/bin/env python3

"""Verify Batch 80 durable Director tombstone and resource cleanup."""

from __future__ import annotations

import json
import os
from pathlib import Path

from playwright.sync_api import Page, sync_playwright


ROOT = Path(__file__).resolve().parents[1]
BASE_URL = os.environ.get("LIBLIB_BASE_URL", "http://localhost:4317")
AUDIT_PATH = (
    ROOT
    / "docs"
    / "research"
    / "liblib-canvas-batch80-2026-08-28"
    / "runtime-audit.json"
)
LOCAL_MODEL_KEY = "liblib-tv-director-local-model-library-v1"


def attach_errors(page: Page) -> tuple[list[str], list[str]]:
    errors: list[str] = []
    phase = ["startup"]
    page.on(
        "console",
        lambda message: errors.append(
            f"{phase[0]}:console:{message.type}:{message.text}"
        )
        if message.type == "error"
        else None,
    )
    page.on(
        "pageerror",
        lambda error: errors.append(
            f"{phase[0]}:pageerror:{error}\n{error.stack or 'stack unavailable'}"
        ),
    )
    page.on(
        "requestfailed",
        lambda request: errors.append(
            f"{phase[0]}:requestfailed:{request.method}:{request.url}:"
            f"{request.failure}"
        ),
    )
    return errors, phase


def wait_for_app(page: Page) -> None:
    page.wait_for_function(
        "() => Boolean(window.__libtv_store && window.__director_store)"
    )
    page.wait_for_selector("[data-libtv-react-flow-host]")


def main() -> None:
    with sync_playwright() as playwright:
        browser = playwright.chromium.launch()
        context = browser.new_context(viewport={"width": 1440, "height": 900})
        page = context.new_page()
        errors, phase = attach_errors(page)
        page.goto(BASE_URL, wait_until="domcontentloaded")
        wait_for_app(page)

        phase[0] = "reset"
        page.evaluate(
            """(modelKey) => {
              for (const key of Object.keys(localStorage)) {
                if (key.startsWith("liblib-tv-director-project-v1:")) {
                  localStorage.removeItem(key);
                }
              }
              localStorage.removeItem(modelKey);
            }""",
            LOCAL_MODEL_KEY,
        )
        page.reload(wait_until="domcontentloaded")
        wait_for_app(page)
        page.evaluate(
            "() => window.__director_store.getState().hydrateLocalModelLibrary()"
        )

        phase[0] = "prepare"
        fixture = page.evaluate(
            """() => {
              const canvasState = window.__libtv_store.getState();
              const canvas = canvasState.canvases.find((item) =>
                item.nodes.some((node) => node.type === "script-execution")
              );
              if (!canvas) throw new Error("Director canvas fixture is missing");
              canvasState.setActiveCanvas(canvas.id);
              const source = canvas.nodes.find(
                (node) => node.type === "script-execution"
              );
              if (!source) throw new Error("Director source fixture is missing");
              canvasState.addNodeAtPosition(
                "script-execution",
                { x: 1800, y: -187 },
                { title: "Batch80 second Director owner" }
              );
              return { canvasId: canvas.id, sourceNodeId: source.id };
            }"""
        )
        page.wait_for_function(
            """(canvasId) => {
              const canvas = window.__libtv_store.getState().canvases.find(
                (item) => item.id === canvasId
              );
              return (canvas?.nodes || []).filter(
                (node) => node.type === "script-execution"
              ).length >= 2;
            }""",
            arg=fixture["canvasId"],
        )
        fixture["secondNodeId"] = page.evaluate(
            """(fixture) => {
              const canvas = window.__libtv_store.getState().canvases.find(
                (item) => item.id === fixture.canvasId
              );
              return canvas?.nodes.find(
                (node) =>
                  node.type === "script-execution" &&
                  node.id !== fixture.sourceNodeId
              )?.id || null;
            }""",
            fixture,
        )
        assert fixture["secondNodeId"], fixture

        local_item = {
            "id": "batch80-shared-local-model",
            "categoryId": "my-models",
            "name": "Batch 80 shared model",
            "fileName": "batch80-shared.obj",
            "dataUrl": "data:text/plain;base64,byBiYXRjaDgwLXNoYXJlZAp2IDAgMCAwCnYgMSAwIDAKdiAwIDEgMApmIDEgMiAzCg==",
            "visual": "box",
            "color": "#7db5d8",
        }
        page.evaluate(
            """(item) =>
              window.__director_store.getState().addLocalModelLibraryItem(item)""",
            local_item,
        )

        owner_a = {
            "route": "libtv",
            "canvasId": fixture["canvasId"],
            "sourceNodeId": fixture["sourceNodeId"],
        }
        owner_b = {
            "route": "libtv",
            "canvasId": fixture["canvasId"],
            "sourceNodeId": fixture["secondNodeId"],
        }

        phase[0] = "author-source-a"
        page.evaluate(
            """(owner) => {
              window.__libtv_ui_store.getState().openDirectorDesk(
                owner.sourceNodeId,
                owner.canvasId
              );
            }""",
            owner_a,
        )
        page.locator("[data-director-workspace]").wait_for(state="visible")
        page.wait_for_function(
            "(owner) => JSON.stringify(window.__director_store.getState().projectOwner) === JSON.stringify(owner)",
            arg=owner_a,
        )
        page.evaluate(
            """(item) => {
              const state = window.__director_store.getState();
              state.addModelLibraryObject(item);
              const object = window.__director_store.getState().authoredObjects.find(
                (candidate) => candidate.libraryAssetId === item.id
              );
              if (!object) throw new Error("source local model object missing");
              state.updateObject(object.id, { name: "Batch80 source A authored" });
              const current = window.__director_store.getState();
              const camera = current.authoredObjects.find(
                (candidate) => candidate.id === current.activeCameraId
              );
              current.addCapture({
                id: "batch80-capture-a",
                dataUrl: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNk+A8AAQUBAScY42YAAAAASUVORK5CYII=",
                cameraId: current.activeCameraId,
                cameraName: camera?.name || "Main camera",
                aspectRatio: current.aspectRatio,
                width: 1280,
                height: 720,
                createdAt: "2026-08-28T12:10:00.000Z",
              });
            }""",
            local_item,
        )
        source_a_before_close = page.evaluate(
            """() => {
              const state = window.__director_store.getState();
              return {
                projectId: state.projectId,
                objectCount: state.authoredObjects.length,
                localResourceRefs: state.authoredObjects.filter(
                  (object) => object.libraryAssetId === "batch80-shared-local-model"
                ).length,
                captureCount: state.captures.length,
                historyPast: state.history.past.length,
              };
            }"""
        )
        page.evaluate(
            """(owner) => {
              window.__director_store.getState().closeSession(owner);
              window.__libtv_ui_store.getState().closeDirectorDesk();
            }""",
            owner_a,
        )
        page.locator("[data-director-workspace]").wait_for(state="hidden")

        phase[0] = "author-source-b"
        page.evaluate(
            """(owner) => {
              window.__libtv_ui_store.getState().openDirectorDesk(
                owner.sourceNodeId,
                owner.canvasId
              );
            }""",
            owner_b,
        )
        page.locator("[data-director-workspace]").wait_for(state="visible")
        page.wait_for_function(
            "(owner) => JSON.stringify(window.__director_store.getState().projectOwner) === JSON.stringify(owner)",
            arg=owner_b,
        )
        page.evaluate(
            """(item) => {
              const state = window.__director_store.getState();
              state.addModelLibraryObject(item);
              const object = window.__director_store.getState().authoredObjects.find(
                (candidate) => candidate.libraryAssetId === item.id
              );
              if (!object) throw new Error("source B local model object missing");
              state.updateObject(object.id, { name: "Batch80 source B authored" });
              state.closeSession();
              window.__libtv_ui_store.getState().closeDirectorDesk();
            }""",
            local_item,
        )
        page.locator("[data-director-workspace]").wait_for(state="hidden")

        authored = page.evaluate(
            """(owners) => {
              const registry = window.__director_project_registry_snapshot();
              return owners.map((owner) => {
                const record = registry.records.find(
                  (candidate) =>
                    JSON.stringify(candidate.identity.owner) === JSON.stringify(owner)
                );
                return record
                  ? {
                      owner: record.identity.owner,
                      projectId: record.identity.projectId,
                      lifecycle: record.lifecycle,
                      localRefs: record.document.resourceRefs
                        .filter((resource) => resource.source === "local")
                        .map((resource) => resource.id),
                      captures: record.memory.captures.length,
                    }
                  : null;
              });
            }""",
            [owner_a, owner_b],
        )
        assert authored[0] and authored[1], authored
        assert authored[0]["projectId"] != authored[1]["projectId"], authored
        assert authored[0]["captures"] == 1, authored
        assert authored[0]["localRefs"] == [local_item["id"]], authored
        assert authored[1]["localRefs"] == [local_item["id"]], authored

        phase[0] = "delete-active-owner"
        page.evaluate(
            """(owner) => {
              window.__libtv_ui_store.getState().openDirectorDesk(
                owner.sourceNodeId,
                owner.canvasId
              );
            }""",
            owner_a,
        )
        page.locator("[data-director-workspace]").wait_for(state="visible")
        page.wait_for_function(
            "(owner) => JSON.stringify(window.__director_store.getState().projectOwner) === JSON.stringify(owner)",
            arg=owner_a,
        )
        page.evaluate(
            """(owner) => {
              window.__libtv_store.getState().removeNode(owner.sourceNodeId);
            }""",
            owner_a,
        )
        page.wait_for_function(
            """() =>
              window.__director_store.getState().projectId === null &&
              window.__director_store.getState().sessionId === null &&
              window.__director_store.getState().localModelLibrary.some(
                (item) => item.id === "batch80-shared-local-model"
              )"""
        )
        page.locator("[data-director-workspace]").wait_for(state="hidden")

        after_active_delete = page.evaluate(
            """(owner) => {
              const registry = window.__director_project_registry_snapshot();
              const persistence = window.__director_project_persistence_snapshot();
              const record = registry.records.find(
                (candidate) =>
                  JSON.stringify(candidate.identity.owner) === JSON.stringify(owner)
              );
              const persisted = persistence.records.find(
                (candidate) =>
                  JSON.stringify(candidate.owner) === JSON.stringify(owner)
              );
              const canvasState = window.__libtv_store.getState();
              const canvas = canvasState.canvases.find(
                (item) => item.id === owner.canvasId
              );
              const history = canvasState.historyByCanvas[owner.canvasId] || {
                past: [],
                future: [],
              };
              return {
                registry: record
                  ? {
                      lifecycle: record.lifecycle,
                      memoryCaptureCount: record.memory.captures.length,
                      generation: record.identity.generation,
                    }
                  : null,
                persistence: persisted
                  ? {
                      status: persisted.status,
                      reason: persisted.reason,
                      generation: persisted.generation,
                    }
                  : null,
                localModelIds: window.__director_store
                  .getState()
                  .localModelLibrary.map((item) => item.id),
                remainingDirectorNodes: (canvas?.nodes || [])
                  .filter((node) => node.type === "script-execution")
                  .map((node) => node.id),
                graphHistoryPast: history.past.length,
                directorHistoryPast: window.__director_store.getState().history.past.length,
                activeSession: registry.activeSession,
              };
            }""",
            owner_a,
        )
        assert after_active_delete["registry"]["lifecycle"] == "TOMBSTONED", after_active_delete
        assert after_active_delete["registry"]["memoryCaptureCount"] == 0, after_active_delete
        assert after_active_delete["persistence"]["status"] == "TOMBSTONED", after_active_delete
        assert local_item["id"] in after_active_delete["localModelIds"], after_active_delete
        assert owner_b["sourceNodeId"] in after_active_delete["remainingDirectorNodes"], after_active_delete
        assert after_active_delete["graphHistoryPast"] >= 1, after_active_delete
        assert after_active_delete["directorHistoryPast"] == 0, after_active_delete
        assert after_active_delete["activeSession"] is None, after_active_delete

        phase[0] = "delete-inactive-owner"
        page.evaluate(
            """(owner) => {
              window.__libtv_store.getState().removeNode(owner.sourceNodeId);
            }""",
            owner_b,
        )
        page.wait_for_function(
            """() =>
              !window.__director_store.getState().localModelLibrary.some(
                (item) => item.id === "batch80-shared-local-model"
              )"""
        )
        after_inactive_delete = page.evaluate(
            """(owner) => {
              const registry = window.__director_project_registry_snapshot();
              const persistence = window.__director_project_persistence_snapshot();
              const record = registry.records.find(
                (candidate) =>
                  JSON.stringify(candidate.identity.owner) === JSON.stringify(owner)
              );
              const persisted = persistence.records.find(
                (candidate) =>
                  JSON.stringify(candidate.owner) === JSON.stringify(owner)
              );
              const rawModel = localStorage.getItem(
                "liblib-tv-director-local-model-library-v1"
              );
              const canvas = window.__libtv_store.getState().canvases.find(
                (item) => item.id === owner.canvasId
              );
              const history =
                window.__libtv_store.getState().historyByCanvas[owner.canvasId] || {
                  past: [],
                  future: [],
                };
              return {
                registry: record
                  ? {
                      lifecycle: record.lifecycle,
                      memoryCaptureCount: record.memory.captures.length,
                    }
                  : null,
                persistence: persisted
                  ? {
                      status: persisted.status,
                      reason: persisted.reason,
                    }
                  : null,
                modelStorage: rawModel,
                remainingNodes: (canvas?.nodes || []).filter(
                  (node) => node.type === "script-execution"
                ).length,
                graphHistoryPast: history.past.length,
              };
            }""",
            owner_b,
        )
        assert after_inactive_delete["registry"]["lifecycle"] == "TOMBSTONED", after_inactive_delete
        assert after_inactive_delete["registry"]["memoryCaptureCount"] == 0, after_inactive_delete
        assert after_inactive_delete["persistence"]["status"] == "TOMBSTONED", after_inactive_delete
        assert after_inactive_delete["modelStorage"] in (None, "[]"), after_inactive_delete
        assert after_inactive_delete["remainingNodes"] == 0, after_inactive_delete
        assert after_inactive_delete["graphHistoryPast"] >= 2, after_inactive_delete

        phase[0] = "reload-reopen-guard"
        page.reload(wait_until="domcontentloaded")
        wait_for_app(page)
        page.evaluate(
            "() => window.__director_store.getState().hydrateLocalModelLibrary()"
        )
        reopen_a = page.evaluate(
            """(owner) =>
              window.__director_store.getState().openSession(owner)""",
            owner_a,
        )
        reopen_b = page.evaluate(
            """(owner) =>
              window.__director_store.getState().openSession(owner)""",
            owner_b,
        )
        assert reopen_a["disposition"] == "REJECTED", reopen_a
        assert reopen_a["reason"] == "PROJECT_TOMBSTONED", reopen_a
        assert reopen_b["disposition"] == "REJECTED", reopen_b
        assert reopen_b["reason"] == "PROJECT_TOMBSTONED", reopen_b
        after_reload = page.evaluate(
            """() => ({
              localModelLibrary: window.__director_store.getState()
                .localModelLibrary,
              persistence: window.__director_project_persistence_snapshot(),
              errors: [],
            })"""
        )
        assert after_reload["localModelLibrary"] == [], after_reload
        assert not errors, errors

        phase[0] = "complete"
        audit = {
            "batch": 80,
            "status": "SCRIPT_RECORDED_PASS",
            "baseUrl": BASE_URL,
            "owners": {
                "sourceA": owner_a,
                "sourceB": owner_b,
                "sourceAProjectId": authored[0]["projectId"],
                "sourceBProjectId": authored[1]["projectId"],
            },
            "activeOwnerCleanup": {
                "tombstoned": True,
                "memoryCaptureArchiveCleared": after_active_delete["registry"][
                    "memoryCaptureCount"
                ]
                == 0,
                "sharedLocalResourceRetained": local_item["id"]
                in after_active_delete["localModelIds"],
                "activeSessionCleared": after_active_delete["activeSession"] is None,
                "directorHistoryUnaffected": after_active_delete["directorHistoryPast"]
                == 0,
            },
            "inactiveOwnerCleanup": {
                "tombstoned": True,
                "unreferencedLocalResourceReleased": after_inactive_delete[
                    "modelStorage"
                ]
                in (None, "[]"),
                "graphHistoryPreserved": after_inactive_delete["graphHistoryPast"]
                >= 2,
            },
            "reload": {
                "bothDurableTombstonesRejectReopen": True,
                "localModelLibraryEmpty": after_reload["localModelLibrary"] == [],
            },
            "errors": errors,
        }
        AUDIT_PATH.write_text(
            json.dumps(audit, ensure_ascii=False, indent=2) + "\n",
            encoding="utf-8",
        )
        assert not errors, errors
        print(json.dumps(audit, ensure_ascii=False, separators=(",", ":")))
        context.close()
        browser.close()


if __name__ == "__main__":
    main()
